### PR TITLE
deepchecks properties + weak segments + label correlation

### PR DESCRIPTION
**Pull request by nirhutnik**
Some comments:

1. should "topic" remain? Takes a lot of time and not sure whether it's generic enough
2. Originally we said we want to do additional-data-label-correlation but I don't think so anymore. Opinions?

### CODE REVIEW

**Review by Nadav-Barak — COMMENTED** [review]
I think we need more tests: 
more unique tests per property to make sure they handle edge cases
maybe even to add one additional big dataset to test

File: deepchecks/nlp/checks/data_integrity/property_label_correlation.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        DeepchecksValueError
+            If the object is not a Dataset instance with a label.
+        """
+        text_data = context.get_data_by_kind(dataset_kind).sample(self.n_samples, random_state=context.random_state)
+
+        df = text_data.properties.join(pd.Series(text_data.label, name='label', index=text_data.index).astype(str))
+
+        df_pps = pps.predictors(df=df, y='label', random_seed=context.random_state,
+                                **self.ppscore_params)
+        s_ppscore = df_pps.set_index('x', drop=True)['ppscore']
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/checks/data_integrity/property_label_correlation.py:85]: pps.predictors is not our code. it is ppscore

File: deepchecks/nlp/checks/data_integrity/property_label_correlation.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            data is a bar graph of the PPS of each feature.
+
+        Raises
+        ------
+        DeepchecksValueError
+            If the object is not a Dataset instance with a label.
+        """
+        text_data = context.get_data_by_kind(dataset_kind).sample(self.n_samples, random_state=context.random_state)
+
+        df = text_data.properties.join(pd.Series(text_data.label, name='label', index=text_data.index).astype(str))
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/checks/data_integrity/property_label_correlation.py:81]: What's the problem with df as a var name of something being a dataframe?

File: deepchecks/nlp/checks/data_integrity/property_label_correlation.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+        df_pps = pps.predictors(df=df, y='label', random_seed=context.random_state,
+                                **self.ppscore_params)
+        s_ppscore = df_pps.set_index('x', drop=True)['ppscore']
+
+        if context.with_display:
+            if text_data.label_map:
+                s_ppscore = s_ppscore.reset_index(text_data.label_map, drop=True)
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/checks/data_integrity/property_label_correlation.py:89]: Nope, you're right, also relevant for result.value, thanks!

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py [hunk 4]
```diff
@@ -132,3 +148,17 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
                       'list of weak segments can be observed in the check result value. '
         return CheckResult({'weak_segments_list': weak_segments, 'avg_score': avg_score, 'scorer_name': scorer.name},
                            display=[display_msg, DisplayMap(display)])
+
+
+class PropertySegmentsPerformance(WeakSegmentsPerformance):
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:153]: Yeah :/ just copy paste?

File: deepchecks/nlp/text_data.py [hunk 5]
```diff
@@ -59,13 +61,22 @@ class TextData:
...
         The task type for the text data. Can be either 'text_classification' or 'token_classification'. Must be set if
         label is provided.
     dataset_name : t.Optional[str] , default: None
         The name of the dataset. If None, the dataset name will be defined when running it within a check.
     index : t.Optional[t.Sequence[int]] , default: None
         The index of the samples. If None, the index is set to np.arange(len(raw_text)).
+    additional_data : t.Optional[pd.DataFrame] , default: None
+        Additional data for the samples. If None, no additional data is set. If a DataFrame is given, it must contain
+        the same number of samples as the raw_text and identical index.
+    properties : t.Optional[Union[pd.DataFrame, str]] , default: None
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:76]: I don't want to as the calculations may be long + requires packages they don't necessarily have

File: deepchecks/nlp/text_data.py [hunk 6]
```diff
@@ -302,6 +330,25 @@ def additional_data(self) -> pd.DataFrame:
         """Return the additional data of for the dataset."""
         return self._additional_data
 
+    def calculate_default_properties(self, include_properties: t.List[str] = None,
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:333]: But it's like the columns/ignore_columns scheme in tabular. WDYT?

File: deepchecks/nlp/utils/text_properties.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+    def language(raw_text: Sequence[str]) -> List[str]:
+        """Return list of strings of language."""
+        return [langdetect.detect(text) for text in raw_text]
+
+    return language
+
+
+def get_sentiment_detection() -> callable:
+    """Return a function that returns the language of a text.
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/utils/text_properties.py:65]: yep. Thanks!

File: deepchecks/nlp/utils/text_properties.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+            all_topics += [class_mapping[i] for i in np.argmax(scores, axis=1)]
+
+        return all_topics
+
+    return topic
+
+
+def _get_default_properties(include_properties: List[str] = None, ignore_properties: List[str] = None):
+    """Return the default properties.
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/utils/text_properties.py:109]: Do you have an idea? 
The problem is with those functions that require specific packages :/

File: deepchecks/nlp/utils/text_properties.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+    Default properties are defined here and not outside the function so not to import all the packages
+    if they are not needed.
+    """
+    ret = [
+        {'name': 'text_length', 'method': text_length, 'output_type': 'numeric'},
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/utils/text_properties.py:115]: Discussed f2f and said that for now we want this

File: deepchecks/nlp/checks/data_integrity/property_label_correlation.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+PLC = t.TypeVar('PLC', bound='PropertyLabelCorrelation')
+
+pps_url = 'https://docs.deepchecks.com/en/stable/checks_gallery/tabular/' \
+          'train_test_validation/plot_feature_label_correlation_change.html'
+pps_html = f'<a href={pps_url} target="_blank">Predictive Power Score</a>'
+
+
+class PropertyLabelCorrelation(SingleDatasetCheck):
+    """Return the PPS (Predictive Power Score) of all features in relation to the label.
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/property_label_correlation.py:34]: features? This docstring should be updated for nlp

File: deepchecks/nlp/checks/data_integrity/property_label_correlation.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        self.n_top_features = n_top_features
+        self.n_samples = n_samples
+
+    def run_logic(self, context: Context, dataset_kind) -> CheckResult:
+        """Run check.
+
+        Returns
+        -------
+        CheckResult
+            value is a dictionary with PPS per feature column.
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/property_label_correlation.py:72]: update docstring as well

File: deepchecks/nlp/checks/data_integrity/property_label_correlation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/checks/data_integrity/property_label_correlation.py:82]: agree that it's clear enough as is

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py [hunk 12]
```diff
@@ -84,11 +86,26 @@ def __init__(self, columns: Union[Hashable, List[Hashable], None] = None,
...
         text_data = context.get_data_by_kind(dataset_kind)
-        if text_data.additional_data is None:
-            raise DeepchecksNotSupportedError(
-                'Weak segments check requires additional data to be available in the text data.')
         text_data = text_data.sample(self.n_samples, random_state=context.random_state)
 
+        if self.segment_by == 'additional_data':
+            if text_data.additional_data is None:
+                raise DeepchecksNotSupportedError(
+                    'Weak segments check requires additional data to be available in the text data.')
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:94]: Agree that this could be a good parameter also for the generic weak segment, but it can be an additional issue.

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py [hunk 13]
```diff
@@ -84,11 +86,26 @@ def __init__(self, columns: Union[Hashable, List[Hashable], None] = None,
...
+
+        elif self.segment_by == 'properties':
+            features = select_from_dataframe(text_data.properties, self.columns, self.ignore_columns)
+            features_name = 'properties'
+        else:
+            raise DeepchecksProcessError(f'Unknown segment_by value: {self.segment_by}')
+
+        # TODO: Don't use cat_features but enable user to give their own datatype / use properties known types.
+        # This is here because Dataset object writes a warning if categorical features are not given
+        cat_features = infer_categorical_features(features)
```
Comment (noamzbr) [inline deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:107]: Is that what we want to have in nlp?

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py [hunk 14]
```diff
@@ -102,11 +119,10 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
             loss_per_sample = [log_loss([y_true], [y_proba], labels=sorted(context.model_classes)) for y_true, y_proba
                                in zip(list(text_data.label), proba_values)]
 
-        additional_data = select_from_dataframe(text_data.additional_data, self.columns, self.ignore_columns)
-        if additional_data.shape[1] < 2:
```
Comment (noamzbr) [inline deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:106]: actually I don't know why. I mean, 1D segmentation is more boring but can still be useful.

File: deepchecks/nlp/datasets/classification/tweet_emotion.py [hunk 15]
```diff
@@ -28,13 +28,45 @@
...
+_PROPERTIES_URL = 'https://ndownloader.figshare.com/files/39460924'
 _PREDICTIONS_URL = 'https://ndownloader.figshare.com/files/39264461'
 
 ASSETS_DIR = pathlib.Path(__file__).absolute().parent.parent / 'assets' / 'tweet_emotion'
 _target = 'label'
 
 
-def load_data(data_format: str = 'TextData', as_train_test: bool = True) -> \
+def load_properties(as_train_test: bool = True) -> t.Union[pd.DataFrame, t.Tuple[pd.DataFrame, pd.DataFrame]]:
+    """Load and return the properties of the tweet_emotion dataset.
```
Comment (noamzbr) [inline deepchecks/nlp/datasets/classification/tweet_emotion.py:39]: Add license for dataset.

File: deepchecks/nlp/text_data.py [hunk 16]
```diff
@@ -59,13 +61,22 @@ class TextData:
         - token_classification label - For token classification the accepted label format is the IOB format or similar
           to it. The Label must be a sequence of sequences of strings or integers, with each sequence corresponding to
           a sample in the tokenized text, and exactly the length of the corresponding tokenized text.
+    label_map : t.Optional[t.Dict[str, int]], default: None
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:64]: Why can't the label map be something internal we do behind the scenes to support pps? Why should the user be involved?

File: deepchecks/nlp/text_data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:76]: +1 to Nir

File: deepchecks/nlp/text_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:333]: I don't think this matters a lot, but agree with Nir that it doesn't have to be viable English

File: deepchecks/nlp/utils/text_properties.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+    NLP module, we import it only when needed.
+    Therefore, the language property is not given directly, but rather through this function, so that the user will only
+    get an ImportError if they try to use the language property.
+    """
+    try:
+        import langdetect  # pylint: disable=import-outside-toplevel
+        from langdetect import DetectorFactory  # pylint: disable=import-outside-toplevel
+        DetectorFactory.seed = 42
+    except ImportError as e:
+        raise ImportError(
```
Comment (noamzbr) [inline deepchecks/nlp/utils/text_properties.py:53]: maybe we want these as warnings? so you'll get the properties you can get, and get warnings about ones that you need to do stuff so you can get them.

File: deepchecks/nlp/utils/text_properties.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+    # Filter by properties or ignore_properties:
+    if include_properties is not None and ignore_properties is not None:
+        raise ValueError('Cannot use properties and ignore_properties parameters together.')
+    elif include_properties is not None:
+        ret = [prop for prop in ret if prop['name'] in include_properties]
+    elif ignore_properties is not None:
+        ret = [prop for prop in ret if prop['name'] not in ignore_properties]
+
+    # Add properties that require additional packages:
```
Comment (noamzbr) [inline deepchecks/nlp/utils/text_properties.py:128]: Not sure why need special code for these properties. Can you explain?

File: makefile [hunk 19]
```diff
@@ -172,8 +172,11 @@ vision-torch-tf-setup: env
 
 	@$(PIP) install -q "tensorflow-hub==0.12.0";
 
+nlp-tests-setup: env
```
Comment (noamzbr) [inline makefile:175]: Why not in nlp req file?

Comment (nirhutnik) [reply makefile:175]: Because I don't want to require packages that are needed only for properties.

File: deepchecks/nlp/utils/text_properties.py (hunk 18 above)
Comment (nirhutnik) [reply deepchecks/nlp/utils/text_properties.py:128]: These are the properties that use outside packages that I didn't want to require users to use. That's why their functions return a function, and need this special treatment in order for the imports to be used only when actually calculating the properties

File: deepchecks/nlp/utils/text_properties.py (hunk 17 above)
Comment (nirhutnik) [reply deepchecks/nlp/utils/text_properties.py:53]: Hm. But then you'll need to rerun it all... Weird as a warning in my opinion. You can just ignore these if you want...

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py (hunk 14 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:106]: What?

File: deepchecks/nlp/datasets/classification/tweet_emotion.py (hunk 15 above)
Comment (nirhutnik) [reply deepchecks/nlp/datasets/classification/tweet_emotion.py:39]: I have no idea where its from... You approved Nadav's PR that merged it 😅

File: deepchecks/nlp/checks/data_integrity/property_label_correlation.py (hunk 10 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/data_integrity/property_label_correlation.py:34]: Done

File: deepchecks/nlp/checks/data_integrity/property_label_correlation.py (hunk 11 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/data_integrity/property_label_correlation.py:72]: Done

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:94]: Fixed

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py (hunk 13 above)
Comment (nirhutnik) [reply deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:107]: Didn't understand the question

File: deepchecks/nlp/text_data.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/nlp/text_data.py:64]: the label map is like in vision, it's a translation of integers to the names of the labels... Didn't understand your question

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py (hunk 13 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:105]: Oh I see, it's for additional data. NVM

File: deepchecks/nlp/utils/text_properties.py (hunk 17 above)
Comment (noamzbr) [reply deepchecks/nlp/utils/text_properties.py:53]: How do you ignore an error?

File: deepchecks/nlp/text_data.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/text_data.py:64]: 1. your wrote `t.Dict[str, int]` which means "name to string"
2. In NLP the label is allowed to be a string, so no need to use label map. The user can just give the labels in their "interpretable" form in the first place, like tabular.

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py (hunk 14 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:106]: why won't we allow features.shape[1] == 1

File: deepchecks/nlp/datasets/classification/tweet_emotion.py (hunk 15 above)
Comment (noamzbr) [reply deepchecks/nlp/datasets/classification/tweet_emotion.py:39]: missed it back then, let's cite the appropriate papers from https://huggingface.co/datasets/tweet_eval#licensing-information

**Review by noamzbr — APPROVED** [review]
